### ISSUE TITLE
Degraded performance compared to `EpiAware` version

### ISSUE THREAD

**Issue by SamuelBrand1** [EpiAware/PrimaryCensored.jl#42]
## Degraded performance

Note that the `cdf(censoreddist,...)` approach used in this note is about x7 _slower_ than the straight function approach used at https://cdcgov.github.io/Rt-without-renewal/dev/getting-started/tutorials/censored-obs/ when run locally. 

This suggests that:
- The extra constructor step is causing problems with the compiler and/or the AD system.
- We should have an ongoing benchmark for this.

_Originally posted by @SamuelBrand1 in https://github.com/EpiAware/PrimaryCensored.jl/issues/39#issuecomment-2419156548_
            
This issue is for resolving the performance downgrade described above.

**Comment by seabbs** [comment]
The additional considerations is that the current example approach assumes independence between pmf calls and so does roughly triple the CDF calls of the `EpiAware` version. 

Part of that is built in to the current approach (and is also how the most generic version of the R package works). I would expect us to be able to solve for common cases with specialisations going forward but I think we should focus on the design patterns of the package before that. 

I agree that we should have some benchmarks here to look at performance and that we should move to outer constructors where possible as this appears to be good practice (and follows `Distributions.jl`)

**Comment by SamuelBrand1** [comment]
Adding context here: the `EpiAware` example _doesn't_ vectorise as described by @seabbs above.

There was a first pass version where we did that, but the current version matches the (naive) number of function calls as the version of the example here.

**Comment by seabbs** [comment]
Yes correct so the issue is one of the things @SamuelBrand1 or the structure of the integral or other

**Comment by seabbs** [comment]
Ping
